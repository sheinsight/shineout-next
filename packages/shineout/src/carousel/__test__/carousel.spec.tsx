import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Carousel from '..';
import mountTest from '../../tests/mountTest';
import { classLengthTest } from '../../tests/structureTest';
import {
  baseTest,
  childrenTest,
  classTest,
  createClassName,
  displayTest,
  snapshotTest,
  styleTest,
} from '../../tests/utils';
import CaroselBase from '../__example__/01-base';
import CarouselIndicator from '../__example__/02-indicator';
import CarouselAnimation from '../__example__/03-animation';

const SO_PREFIX = 'carousel';
const originClasses = [
  'wrapper',
  'slider',
  'item',
  'indicatorWrapper',
  'indicator',
  'arrowWrapper',
  'arrowItem',
];
const originItemClasses = [
  'animationSlide',
  'directionStop',
  'itemCurrent',
  'indicatorCenter',
  'indicatorTypeCircle',
  'indicatorTypeNumber',
  'indicatorTypeLine',
  'indicatorTypeSlider',
  'arrowHover',
  'arrowLeft',
  'arrowRight',
  'indicatorActive',
  'indicatorLeft',
  'indicatorRight',
  'indicatorOuter',
  'animationSlideY',
  'animationFade',
];
const {
  wrapper,
  animationSlide,
  directionStop,
  slider,
  item,
  itemCurrent,
  indicatorWrapper,
  indicatorCenter,
  indicatorTypeCircle,
  indicatorTypeNumber,
  indicatorTypeLine,
  indicatorTypeSlider,
  indicator,
  arrowWrapper,
  arrowHover,
  arrowItem,
  arrowLeft,
  arrowRight,
  indicatorActive,
  indicatorLeft,
  indicatorRight,
  indicatorOuter,
  animationSlideY,
  animationFade,
} = createClassName(SO_PREFIX, originClasses, originItemClasses);

const images = [
  'https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-01.png',
  'https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-02.png',
  'https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-03.png',
  'https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-04.png',
];

afterEach(cleanup);

const CarouselType = (props: any) => (
  <Carousel {...props}>
    {images.map((src) => (
      <img key={src} src={src} />
    ))}
  </Carousel>
);

mountTest(<CarouselType />);
describe('Carousel[Base]', () => {
  displayTest(Carousel, 'ShineoutCarousel');
  baseTest(Carousel, wrapper, { backgroundColor: 'red' }, 'background-color: red; height: auto;');
  childrenTest(Carousel, wrapper);
  snapshotTest(<CaroselBase />);
  snapshotTest(<CarouselIndicator />, 'about indicator');
  snapshotTest(<CarouselAnimation />, 'about animation');
  test('should render default', () => {
    const { container } = render(<CarouselType />);
    const carouselWrapper = container.querySelector(wrapper)!;
    classTest(carouselWrapper, animationSlide);
    classTest(carouselWrapper, directionStop);
    styleTest(carouselWrapper, 'height: auto;');
    const carouselSlider = carouselWrapper.querySelector(slider)!;
    const carouselItems = carouselSlider.querySelectorAll(item);
    classLengthTest(carouselSlider, item, images.length);
    carouselItems.forEach((item) => {
      classLengthTest(item, 'img', 1);
    });
    classTest(carouselItems[0], itemCurrent);
    const carouselIndicatorWrapper = carouselWrapper.querySelector(indicatorWrapper)!;
    classTest(carouselIndicatorWrapper, indicatorCenter);
    classTest(carouselIndicatorWrapper, indicatorTypeCircle);
    classLengthTest(carouselIndicatorWrapper, indicator, images.length);
    const indicators = carouselIndicatorWrapper.querySelectorAll(indicator);
    classTest(indicators[0], indicatorActive);
  });
  test('should render when click indicator', () => {
    const { container } = render(<CarouselType />);
    const indicators = container.querySelectorAll(indicator);
    const carouselItems = container.querySelectorAll(item);
    classTest(carouselItems[0], itemCurrent);
    classTest(indicators[0], indicatorActive);
    fireEvent.click(indicators[1]);
    classTest(carouselItems[1], itemCurrent);
    classTest(indicators[1], indicatorActive);
    fireEvent.click(indicators[1]);
    classTest(carouselItems[1], itemCurrent);
    classTest(indicators[1], indicatorActive);
  });
  test('should render when set onMove', () => {
    const moveFn = jest.fn();
    const { container } = render(<CarouselType onMove={moveFn} />);
    const indicators = container.querySelectorAll(indicator);
    fireEvent.click(indicators[1]);
    expect(moveFn.mock.calls.length).toBe(1);
  });
  test('should render when set interval > 0', async () => {
    jest.useFakeTimers();
    const { container } = render(
      <Carousel interval={5000}>
        <div />
        <div />
      </Carousel>,
    );
    const indicators = container.querySelectorAll(indicator);
    const carouselItems = container.querySelectorAll(item);
    classTest(carouselItems[0], itemCurrent);
    classTest(indicators[0], indicatorActive);
    await waitFor(
      () => {
        jest.advanceTimersByTime(5000);
      },
      { timeout: 6000 },
    );

    await waitFor(() => {
      classTest(carouselItems[1], itemCurrent);
      classTest(indicators[1], indicatorActive);
    });
    await waitFor(
      () => {
        jest.advanceTimersByTime(5000);
      },
      { timeout: 6000 },
    );
    await waitFor(() => {
      classTest(carouselItems[0], itemCurrent);
      classTest(indicators[0], indicatorActive);
    });
    fireEvent.mouseEnter(container.querySelector(wrapper)!);
    await waitFor(
      () => {
        jest.advanceTimersByTime(5000);
      },
      { timeout: 6000 },
    );
    await waitFor(() => {
      classTest(carouselItems[0], itemCurrent);
      classTest(indicators[0], indicatorActive);
    });
    fireEvent.mouseLeave(container.querySelector(wrapper)!);
    await waitFor(
      () => {
        jest.advanceTimersByTime(5000);
      },
      { timeout: 6000 },
    );
    await waitFor(() => {
      classTest(carouselItems[1], itemCurrent);
      classTest(indicators[1], indicatorActive);
    });
  });
});
describe('Carousel[Type]', () => {
  test.each(['hover', 'always'])('should render when set showArrow is %s', (type) => {
    const { container } = render(<CarouselType showArrow={type} />);
    const carouselArrowWrapper = container.querySelector(arrowWrapper)!;
    const arrowItems = carouselArrowWrapper.querySelectorAll(arrowItem);
    classLengthTest(carouselArrowWrapper, arrowItem, 2);
    arrowItems.forEach((item) => {
      classLengthTest(item, 'svg', 1);
    });
    classTest(arrowItems[0], arrowLeft);
    classTest(arrowItems[1], arrowRight);
    if (type === 'always') return;
    classTest(carouselArrowWrapper, arrowHover);
  });
  test('should render when set arrowClassName', () => {
    const { container } = render(<CarouselType showArrow='always' arrowClassName='demo' />);
    classTest(container.querySelector(arrowWrapper)!, 'demo');
  });
  test('should render when click arrow', () => {
    const { container } = render(<CarouselType showArrow='always' />);
    const arrowItems = container.querySelectorAll(arrowItem);
    const indicators = container.querySelectorAll(indicator);
    const carouselItems = container.querySelectorAll(item);
    classTest(carouselItems[0], itemCurrent);
    classTest(indicators[0], indicatorActive);
    fireEvent.click(arrowItems[1]);
    classTest(carouselItems[1], itemCurrent);
    classTest(indicators[1], indicatorActive);
    fireEvent.click(arrowItems[0]);
    classTest(carouselItems[0], itemCurrent);
    classTest(indicators[0], indicatorActive);
    fireEvent.click(arrowItems[0]);
    classTest(carouselItems[3], itemCurrent);
    classTest(indicators[3], indicatorActive);
    fireEvent.click(arrowItems[1]);
    classTest(carouselItems[0], itemCurrent);
    classTest(indicators[0], indicatorActive);
    fireEvent.click(arrowItems[1]);
    classTest(carouselItems[1], itemCurrent);
    classTest(indicators[1], indicatorActive);
  });
  const positionClassNameMap: { [key: string]: string } = {
    left: indicatorLeft,
    right: indicatorRight,
    outer: indicatorOuter,
  };
  test.each(['left', 'right', 'outer'])(
    'should render when set indicatorPosition is %s',
    (type) => {
      const { container } = render(<CarouselType indicatorPosition={type} />);
      classTest(container.querySelector(indicatorWrapper)!, positionClassNameMap[type]);
    },
  );
  const animationClassNameMap: { [key: string]: string } = {
    'slide-y': animationSlideY,
    fade: animationFade,
  };
  test.each(['slide-y', 'fade'])('should render when set animation is %s', (type) => {
    const { container } = render(<CarouselType animation={type} />);
    classTest(container.querySelector(wrapper)!, animationClassNameMap[type]);
  });
  const indicatorTypeClassNameMap: { [key: string]: string } = {
    line: indicatorTypeLine,
    number: indicatorTypeNumber,
    slider: indicatorTypeSlider,
  };
  test.each(['line', 'number', 'slider'])('should render when set indicatorType is %s', (type) => {
    const { container } = render(<CarouselType indicatorType={type} />);
    classTest(container.querySelector(indicatorWrapper)!, indicatorTypeClassNameMap[type]);
  });
  test('should render when set indicatorType is function', () => {
    const { container } = render(
      <CarouselType
        indicatorType={() => (
          <>
            {images.map((item, index) => (
              <div className='demo' key={index}>
                {index}
              </div>
            ))}
          </>
        )}
      />,
    );
    classLengthTest(container.querySelector(indicatorWrapper)!, '.demo', 4);
  });
});
