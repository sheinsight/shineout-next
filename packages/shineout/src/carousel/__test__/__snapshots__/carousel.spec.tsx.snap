// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel[Base] should render correctly  1`] = `
<div
  class="so-carousel-wrapper so-carousel-animation-slide so-carousel-direction-stop"
  style="width: 600px; height: auto;"
>
  <div
    class="so-carousel-slider"
    style="height: 280px;"
  >
    <div
      class="so-carousel-item so-carousel-item-current"
    >
      <img
        src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-01.png"
      />
    </div>
    <div
      class="so-carousel-item"
    >
      <img
        src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-02.png"
      />
    </div>
    <div
      class="so-carousel-item"
    >
      <img
        src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-03.png"
      />
    </div>
    <div
      class="so-carousel-item"
    >
      <img
        src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-04.png"
      />
    </div>
  </div>
  <div
    class="so-carousel-indicator-wrapper so-carousel-indicator-center so-carousel-indicator-type-circle"
  >
    <div
      class="so-carousel-indicator so-carousel-indicator-active"
    />
    <div
      class="so-carousel-indicator"
    />
    <div
      class="so-carousel-indicator"
    />
    <div
      class="so-carousel-indicator"
    />
  </div>
  <div
    class="so-carousel-arrow-hover so-carousel-arrow-wrapper"
  >
    <div
      class="so-carousel-arrow-left so-carousel-arrow-item"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.29289 12.7071C6.93241 12.3466 6.90468 11.7794 7.2097 11.3871L7.29289 11.2929L13.2929 5.29289C13.6834 4.90237 14.3166 4.90237 14.7071 5.29289C15.0676 5.65338 15.0953 6.22061 14.7903 6.6129L14.7071 6.70711L10.1221 11.2928C9.7317 11.6835 9.7317 12.3165 10.1221 12.7072L14.7071 17.2929C15.0676 17.6534 15.0953 18.2206 14.7903 18.6129L14.7071 18.7071C14.3466 19.0676 13.7794 19.0953 13.3871 18.7903L13.2929 18.7071L7.29289 12.7071Z"
        />
      </svg>
    </div>
    <div
      class="so-carousel-arrow-right so-carousel-arrow-item"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.7071 12.7071C17.0676 12.3466 17.0953 11.7794 16.7903 11.3871L16.7071 11.2929L10.7071 5.29289C10.3166 4.90237 9.68342 4.90237 9.29289 5.29289C8.93241 5.65338 8.90468 6.22061 9.2097 6.6129L9.29289 6.70711L13.8779 11.2928C14.2683 11.6835 14.2683 12.3165 13.8779 12.7072L9.29289 17.2929C8.93241 17.6534 8.90468 18.2206 9.2097 18.6129L9.29289 18.7071C9.65338 19.0676 10.2206 19.0953 10.6129 18.7903L10.7071 18.7071L16.7071 12.7071Z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Carousel[Base] should render correctly about animation 1`] = `
<div>
  <div
    style="display: flex; margin-bottom: 16px; align-items: center;"
  >
    <div
      style="width: 150px;"
    >
      animation type:
    </div>
    <div
      class="so-radio-group"
    >
      <div
        class="so-radio-wrapper so-radio-wrapper-checked"
      >
        <input
          checked=""
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12ZM1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23C5.92487 23 1 18.0751 1 12ZM12 7C9.23858 7 7 9.23858 7 12C7 14.7614 9.23858 17 12 17C14.7614 17 17 14.7614 17 12C17 9.23858 14.7614 7 12 7Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          slide
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          slide-y
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          fade
        </span>
      </div>
    </div>
  </div>
  <div
    class="so-carousel-wrapper so-carousel-animation-slide so-carousel-direction-stop"
    style="width: 600px; height: auto;"
  >
    <div
      class="so-carousel-slider"
      style="height: 280px;"
    >
      <div
        class="so-carousel-item so-carousel-item-current"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-01.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-02.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-03.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-04.png"
        />
      </div>
    </div>
    <div
      class="so-carousel-indicator-wrapper so-carousel-indicator-center so-carousel-indicator-type-circle"
    >
      <div
        class="so-carousel-indicator so-carousel-indicator-active"
      />
      <div
        class="so-carousel-indicator"
      />
      <div
        class="so-carousel-indicator"
      />
      <div
        class="so-carousel-indicator"
      />
    </div>
  </div>
</div>
`;

exports[`Carousel[Base] should render correctly about indicator 1`] = `
<div>
  <div
    style="display: flex; margin-bottom: 16px; align-items: center;"
  >
    <div
      style="width: 150px;"
    >
      indicator position:
    </div>
    <div
      class="so-radio-group"
    >
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          left
        </span>
      </div>
      <div
        class="so-radio-wrapper so-radio-wrapper-checked"
      >
        <input
          checked=""
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12ZM1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23C5.92487 23 1 18.0751 1 12ZM12 7C9.23858 7 7 9.23858 7 12C7 14.7614 9.23858 17 12 17C14.7614 17 17 14.7614 17 12C17 9.23858 14.7614 7 12 7Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          center
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          right
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          outer
        </span>
      </div>
    </div>
  </div>
  <div
    style="display: flex; margin-bottom: 16px; align-items: center;"
  >
    <div
      style="width: 150px;"
    >
      indicator type:
    </div>
    <div
      class="so-radio-group"
    >
      <div
        class="so-radio-wrapper so-radio-wrapper-checked"
      >
        <input
          checked=""
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12ZM1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23C5.92487 23 1 18.0751 1 12ZM12 7C9.23858 7 7 9.23858 7 12C7 14.7614 9.23858 17 12 17C14.7614 17 17 14.7614 17 12C17 9.23858 14.7614 7 12 7Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          circle
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          line
        </span>
      </div>
      <div
        class="so-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="so-radio-indicator-wrapper"
        >
          <span
            class="so-radio-indicator"
            tabindex="1"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                clip-rule="evenodd"
                d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 1C5.92487 1 1 5.92487 1 12C1 18.0751 5.92487 23 12 23C18.0751 23 23 18.0751 23 12C23 5.92487 18.0751 1 12 1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
        <span
          class="so-radio-desc"
        >
          number
        </span>
      </div>
    </div>
  </div>
  <div
    class="so-carousel-wrapper so-carousel-animation-slide so-carousel-direction-stop"
    style="width: 600px; height: auto;"
  >
    <div
      class="so-carousel-slider"
      style="height: 280px;"
    >
      <div
        class="so-carousel-item so-carousel-item-current"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-01.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-02.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-03.png"
        />
      </div>
      <div
        class="so-carousel-item"
      >
        <img
          src="https://raw.githubusercontent.com/sheinsight/shineout-static/main/shineout-next/images/image/s-04.png"
        />
      </div>
    </div>
    <div
      class="so-carousel-indicator-wrapper so-carousel-indicator-center so-carousel-indicator-type-circle"
    >
      <div
        class="so-carousel-indicator so-carousel-indicator-active"
      />
      <div
        class="so-carousel-indicator"
      />
      <div
        class="so-carousel-indicator"
      />
      <div
        class="so-carousel-indicator"
      />
    </div>
  </div>
</div>
`;
